Follow real AtlanticPassAPCZoneControlZone physical mode on Overkiz (HEAT, COOL or HEAT_COOL)

Tronix117 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The following PR: #110510 introduced simple support for ZoneControlZone for Atlantic devices.

Temperature change when AUTO mode is enable, was changing cooling as well as heating temperature, this could be lead to unexpected switch between heating/cooling, as well as higher power consumptions.

This PR fixes that and simplifies the logic by always using a single mode for the subsidiary Zones (the mode depends of the physical configuration of the zone, one of  HEAT_COOL, HEAT or COOL).

![Screenshot](https://github.com/home-assistant/core/assets/211541/5ac93c92-be26-4e1c-a9ff-074e157333c9)

With this fix, hvac_action is set correctly, reflecting the real heating / cooling operating state.
Now, min/max temperature are also in synced with the device configuration.

Like the introduction of the support for those device is included in the 2024.03 release, it would be great to also introduce this fix/refactor. (It's better for the users waiting for the support of those device, since it would not provoke a change of interface in future release)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #112602
- This PR is related to issue: #89746 and #108211 )
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Tronix117 committed:
Support HEAT_COOL when mode is Auto on overkiz AtlanticPassAPCZoneControlZone

Tronix117 commented on homeassistant/components/overkiz/climate_entities/atlantic_pass_apc_zone_control_zone.py:
```diff
@@ -60,48 +116,83 @@ def is_using_derogated_temperature_fallback(self) -> bool:
-        if self._attr_hvac_modes != [zone_control_hvac_mode, HVACMode.OFF]:
-            self._attr_hvac_modes = [zone_control_hvac_mode, HVACMode.OFF]
-            self.async_write_ha_state()
+        self.sync_attrs_from_mode(selectable_hvac_mode)
```
@MartinHjelmare thank you for this feedback, I was crying when I wrote it (main reason why it's a Draft PR).
With the HEAT_COOL (for every main device mode), this may not be needed anymore.

We however loose a bit of control on the device because:
if I enable HEAT_COOL (on subsidiaries devices)for other mode than AUTO (on main device), it seams cooling temperature needs to be strictly higher than heating temperature => so it means, if I want to set heating to 21 and cooling to 20 (for whatever reason), and switching between HEATING and COOLING with an automation, I can not do that.
However it may still makes some sense, because then I'll be able to always choose cooling AND heating temperature whatever the mode (with the only limitation being the cooling_temp > heating_temp)

I'll go this way, I'll handle that asap.

Tronix117 committed:
Refactor ZoneControlZone to simplify usic by only using a single hvac mode

Tronix117 commented:
@iMicknl @MartinHjelmare I would like some feedback on those changes.

I simplified the logic by considering Zones having only one mode (depending on the zone device capabilities) in addition to `OFF`. The mode can be `HEAT`, `COOL` or `HEAT_COOL` and does not change through the life of the device.

It should prevent some unexpected behavio when switching the main device between AUTO, COOLING or HEATING.

I still preserve the `is_using_derogated_temperature_fallback` for backward compatibility purpose, even if I don't have yet the knowledge of a device with this Widget / ControllableName, falling into this consideration.

Tronix117 committed:
Fix linting issues

MartinHjelmare commented:
Why did you change the target branch to rc?

Tronix117 commented:
This PR introduces some fixes for the #110510 that has been added to the release, as well as a better way of handling those newly supported devices.

I think it would be great to have it on this release, so that the users waiting for this implementations will directly have this way of controlling the device and it will not change in a future release (cf. #89746 and #108211 )

However I was not sure about the process here, so if you think it's too much changes and it should be postpone until next month, I can switch it back to dev.

jpbede commented:
> I think it would be great to have it on this release, so that the users waiting for this implementations will directly have this way of controlling the device and it will not change in a future release (cf. https://github.com/home-assistant/core/issues/89746 and https://github.com/home-assistant/core/issues/108211 )

Always target the `dev` branch and make sure in your PR description that this is a fix and should be included in the beta release/patch release. The core team will than take care an tag the PR for the right milestone.

Tronix117 commented:
> > I think it would be great to have it on this release, so that the users waiting for this implementations will directly have this way of controlling the device and it will not change in a future release (cf. #89746 and #108211 )
> 
> Always target the `dev` branch and make sure in your PR description that this is a fix and should be included in the beta release/patch release. The core team will than take care and tag the PR for the right milestone.

Sorry about that, and thanks for your answer. I changed it back, and added some notes to the description.

Tronix117 committed:
Makes more sense to use halves there

MartinHjelmare commented on homeassistant/components/overkiz/climate_entities/atlantic_pass_apc_zone_control_zone.py:
```diff
@@ -251,3 +417,51 @@ async def async_refresh_modes(self) -> None:
+                ),
+            )
+
+        return self._attr_min_temp
```
Where is this attribute set?

Tronix117 replied:
Indeed, it was a mistake, I used `super` instead.

MartinHjelmare commented on homeassistant/components/overkiz/climate_entities/atlantic_pass_apc_zone_control_zone.py:
```diff
@@ -251,3 +417,51 @@ async def async_refresh_modes(self) -> None:
+                ),
+            )
+
+        return self._attr_max_temp
```
And this?

MartinHjelmare reviewed (changes requested):
There's a merge conflict.

@iMicknl please review.

iMicknl commented:
@Tronix117 can you rebase on latest dev? I will have a look at your PR afterwards.

Tronix117 committed:
Fix PR feedback

Tronix117 commented on homeassistant/components/overkiz/climate_entities/atlantic_pass_apc_zone_control_zone.py:
```diff
@@ -24,32 +32,127 @@
 
 PRESET_MODES_TO_OVERKIZ = {v: k for k, v in OVERKIZ_MODE_TO_PRESET_MODES.items()}
 
-TEMPERATURE_ZONECONTROL_DEVICE_INDEX = 20
```
Please note this is a regression from #114006 (a test value that should definitely not have been commited after the last rebase)

The correct value has been included in this PR (L54)

iMicknl reviewed (approved):
Thanks a lot, @Tronix117! Very complex device unfortunately, but great to see it well-mapped to HA entity system.
